Escape < and > when serializing HTML attribute values

annevk opened the pull request:
#### b3e11f0a00aa4620e1a3c7c26402913280b6bdea
<pre>
Escape &lt; and &gt; when serializing HTML attribute values
<a href="https://bugs.webkit.org/show_bug.cgi?id=292432">https://bugs.webkit.org/show_bug.cgi?id=292432</a>
<a href="https://rdar.apple.com/150520333">rdar://150520333</a>

Reviewed by Ryosuke Niwa and Sam Weinig.

Implement <a href="https://github.com/whatwg/html/pull/6362">https://github.com/whatwg/html/pull/6362</a> and add a preference
in case we find out we have to disable it in certain cases.

The preference does not impact the recently added getHTML() method on
Element and ShadowRoot as that should be new enough to not matter.

Add new test from: <a href="https://github.com/web-platform-tests/wpt/pull/27501">https://github.com/web-platform-tests/wpt/pull/27501</a>

Canonical link: <a href="https://commits.webkit.org/295149@main">https://commits.webkit.org/295149@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b14c70221dc4e6e94303b81250b151b66d22879f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/103077 "3 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/22752 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/13071 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/108245 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/53720 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/105116 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/23083 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/31252 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/78368 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/35311 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/106083 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/17891 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/93007 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/58701 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/17733 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/11048 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/53075 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/95752 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/87543 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/11113 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/110618 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/101688 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/30214 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/22272 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/87356 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/30579 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/89207 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/86981 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/31830 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/9545 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/24532 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/16921 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/30141 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/35463 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/125321 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/29949 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/34763 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/33276 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/31511 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/editing/MarkupAccumulator.cpp:
```diff
@@ -448,7 +448,8 @@ StringBuilder MarkupAccumulator::takeMarkup()
 {
     appendCharactersReplacingEntities(result, attribute, 0, attribute.length(),
-        inXMLFragmentSerialization() ? EntityMaskInAttributeValue : EntityMaskInHTMLAttributeValue);
+        m_serializationSyntax == SerializationSyntax::XML ? EntityMaskInAttributeValue :
```
This is hard to read. Can we use a lambda & switch instead like this:
```cpp
auto entityMaskMode = [&] {
    switch (m_serializationSyntax) {
    case SerializationSyntax::XML:
        return EntityMaskInAttributeValue;
    case SerializationSyntax::HTML:
        return EntityMaskInHTMLAttributeValue;
    case SerializationSyntax::HTMLLegacyAttributeValue:
        return EntityMaskInHTMLLegacyAttributeValue;
    }
}();
appendCharactersReplacingEntities(result, attribute, 0, attribute.length(), entityMaskMode);
```

annevk committed:
Escape < and > when serializing HTML attribute values
https://bugs.webkit.org/show_bug.cgi?id=292432
rdar://150520333

Reviewed by Ryosuke Niwa and Sam Weinig.

Implement https://github.com/whatwg/html/pull/6362 and add a preference
in case we find out we have to disable it in certain cases.

The preference does not impact the recently added getHTML() method on
Element and ShadowRoot as that should be new enough to not matter.

Add new test from: https://github.com/web-platform-tests/wpt/pull/27501

Canonical link: https://commits.webkit.org/295149@main
